### ISSUE TITLE
gitignore pattern *~ (Emacs backup files) doesn't work 

### ISSUE THREAD

**Issue by oliversturm** [npm/npm#18365]
#### I'm opening this issue because:

  - [x] npm is doing something I don't understand.

#### What's going wrong?

My `.gitignore` contains the pattern `*~` to exclude Emacs backup files. This works correctly. I copied `.gitignore` to `.npmignore` because I needed to add exclusions for some folders in my project. When I ran `npm publish` and then test-installed the resulting package, I found that some files ending in `~` were included in the package.

I found that my source file hierarchy currently has 23 files ending in `~`. Out of these, 5 were included in the package. The five files appear in random locations in the hierarchy. I have no idea why they were included.

Edit: I meant *random locations* as in "random in the source hierarchy". I ran `npm publish` three times now on this package, and it includes the same five `*~` files every time.

#### How can the CLI team reproduce the problem?

I'm not sure. To help the effort, I have created a zip file of my local working directory, including the `*~` files that are not in my github repo. The zip file is here: https://www.dropbox.com/s/6dktipsr6rkcazq/generator-devextreme.zip?dl=0

The package I published is this: https://www.npmjs.com/package/generator-devextreme
The files ending in `~` can be seen in the tar file: https://registry.npmjs.org/generator-devextreme/-/generator-devextreme-1.0.3.tgz

### supporting information:

 - `npm -v` prints: 5.3.0
 - `node -v` prints: v8.4.0
 - `npm config get registry` prints: https://registry.npmjs.org/
 - Windows, OS X/macOS, or Linux?: Linux

**Comment by oliversturm** [comment]
I have an update. Some of the folders of my source hierarchy contained `.gitignore` files, because these were used as templates for the generator. When testing the generator, I found that these files had been renamed to `.npmignore` in the tar archive for the published package. Does that make any sense? I don't know - but it broke my generator. I'm sure there's a reason for this renaming, but I find it very confusing by itself.

Anyway, so I changed my templating logic and now the `.gitignore` files in my templates are only called `gitignore`. My generator now works correctly. And when I double-checked the files installed with the package (and the tar file), I found that now **all** the `*~` files are included. 

So, to summarize:

* My previous structure, which included a bunch of extra `.gitignore` files, obviously confused `npm publish` to no end
* In my current structure, which has a `.gitignore` and a `.npmignore` on the top level, both of which exclude `*~` files, all existing `*~` files are nevertheless packed up and delivered by `npm publish`.

**Comment by oliversturm** [comment]
Another update. I have tested some more and found that the `.npmignore` in my project is unnecessary. So I'm down to one basic `.gitignore`. I also did a simple reproducible test now and I'm ready to call this behavior a bug. I'm going to change the title of this issue. Here are the instructions to reproduce:

```shell
mkdir testnpm
cd testnpm
git init
npm init -y
mkdir dist
touch dist/foo.js dist/foo.js~
echo '*~' > .gitignore
git add --all
git status -s
```

At this point, git will show three files added: `.gitignore`, `package.json` and `dist/foo.js`. Note that `dist/foo.js~` has been correctly excluded by the pattern in `.gitignore`.

Now edit `package.json` and insert this:

```js
"files": [
  "dist"
]
```

Now run these commands:

```shell
npm pack
tar tzvf testnpm-1.0.0.tgz
```

You will see three files in the packed archive: `package/package.json`, `package/dist/foo.js` and `package/dist/foo.js~`. **The file `dist/foo.js~` has not been excluded, even though it matches the pattern `*~` in `.gitignore`.**
